src/Language/PureScript/Errors.hs

```diff
@@ -800,14 +801,40 @@ prettyPrintSingleError (PPEOptions codeColor full level showDocs relPath) e = fl
                       zipWith (Box.<>)
                       (Box.alignHoriz Box.left longestId <$> idBoxes)
                       tyBoxes
+              (xs, ys) = partition resultVariablesInConstraints idents
+              makeCodeBox = markCodeBox . indent . formatTS . unzip . take maxTSResults
             in [ line "You could substitute the hole with one of these values:"
-               , markCodeBox (indent (formatTS (unzip (take maxTSResults idents))))
+               , makeCodeBox xs
+               , line "You may also want these, but these may be less useful:"
+               , makeCodeBox ys
                ]
           _ -> []
       in
         paras $ [ line $ "Hole '" <> markCode name <> "' has the inferred type "
                 , markCodeBox (indent (typeAsBox ty))
                 ] ++ tsResult ++ renderContext ctx
+          where
+            resultVariablesInConstraints (_, t) =
+              not . S.null $ S.intersection constrained results
```

Comment:
This test seems wrong to me. If there are no constraints, then the intersection is empty, so every function with no constraints goes in the useless bucket. I think the test should be something like _every type variable in the output is either constrained or mentioned in some input_.